frame/vesting/src/lib.rs

```diff
@@ -302,121 +407,322 @@ pub mod pallet {
+			Self::do_vested_transfer(source, target, schedule)
+		}
 
-			let target = T::Lookup::lookup(target)?;
-			let source = T::Lookup::lookup(source)?;
-			ensure!(!Vesting::<T>::contains_key(&target), Error::<T>::ExistingVestingSchedule);
+		/// Merge two vesting schedules together, creating a new vesting schedule that unlocks over
+		/// highest possible start and end blocks. If both schedules have already started the current
+		/// block will be used as the schedule start; with the caveat that if one schedule is finished
+		/// by the current block, the other will be treated as the new merged schedule, unmodified.
+		///
+		/// NOTE: If `schedule1_index == schedule2_index` this is a no-op.
+		/// NOTE: This will unlock all schedules through the current block prior to merging.
+		/// NOTE: If both schedules have ended by the current block, no new schedule will be created
+		/// and both will be removed.
+		///
+		/// Merged schedule attributes:
+		/// - `starting_block`: `MAX(schedule1.starting_block, scheduled2.starting_block, current_block)`.
+		/// - `ending_block`: `MAX(schedule1.ending_block, schedule2.ending_block)`.
+		/// - `locked`: `schedule1.locked_at(current_block) + schedule2.locked_at(current_block)`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schedule1_index`: index of the first schedule to merge.
+		/// - `schedule2_index`: index of the second schedule to merge.
+		#[pallet::weight(
+			T::WeightInfo::not_unlocking_merge_schedules(MaxLocksOf::<T>::get())
+			.max(T::WeightInfo::unlocking_merge_schedules(MaxLocksOf::<T>::get()))
+		)]
+		pub fn merge_schedules(
+			origin: OriginFor<T>,
+			schedule1_index: u32,
+			schedule2_index: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			if schedule1_index == schedule2_index {
+				return Ok(());
+			};
+			let schedule1_index = schedule1_index as usize;
+			let schedule2_index = schedule2_index as usize;
+			let vesting = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
 
-			T::Currency::transfer(&source, &target, schedule.locked, ExistenceRequirement::AllowDeath)?;
+			// The schedule index is based off of the schedule ordering prior to filtering out any
+			// schedules that may be ending at this block.
+			let schedule1 = *vesting.get(schedule1_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let schedule2 = *vesting.get(schedule2_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let filter = Filter::Two(schedule1_index, schedule2_index);
 
-			Self::add_vesting_schedule(&target, schedule.locked, schedule.per_block, schedule.starting_block)
-				.expect("user does not have an existing vesting schedule; q.e.d.");
+			// The length of vesting decreases by 2 here since we filter out 2 schedules. Thus we know
+			// below that we can safely insert the new merged schedule.
+			let maybe_vesting = Self::update_lock_and_schedules(who.clone(), vesting, filter);
```

Comment:
I won't raise it as a series issue, but I think `enum Filter` is not a worthwhile abstraction to keep around. You can just change `update_lock_and_schedules` `remove_schedules(to_remove: impl IntoIterator<Item = index>)` and internally, instead of using `should_remove`, remove the intersection of the full vested iterator and `to_remove`. 

Also, `update_lock_and_schedules` is now doing two things: Update the vest to remove any schedule that is expired already, AND remove `filter`. This is clearly saying they should be two separate functions, and chained once needed.

Reply:
> Also, update_lock_and_schedules is now doing two things

Agreed, `update_lock_and_schedules` is a convoluted function with storage writes that is trying to do too many things. I like your proposed solution of creating chain-able functions.

> I won't raise it as a series issue, but I think enum Filter is not a worthwhile abstraction to keep around

I am not convinced it is either, but my plan right now is to rename it `SchedulesAction` with the variants `Remove(idx)`, `Merge(idx1, idx2)`, `Nothing`. An iterator on the other hand indicates we could be expecting an arbitrary amount of schedules to remove. I will reassess once I refactor

Reply:
Here is the commitment with updates 4aaa158